### PR TITLE
Plugin packages

### PR DESCRIPTION
**Pull request by liorblob**
## Status
On Hold(Waiting for the plugin)

Adjust the `package_creator.py` script to work with the new plugin mechanism.

**DO NOT MERGE**

### CODE REVIEW

File: package_creator.py [hunk 1]
```diff
@@ -54,10 +59,11 @@ def merge_script_package_to_yml(package_path, dir_name, dest_path=""):
 
     yml_text, script_path = insert_script_to_yml(package_path, script_type, yml_text, dir_name, yml_data)
     yml_text, image_path = insert_image_to_yml(dir_name, package_path, yml_data, yml_text)
+    yml_text, desc_path = insert_description_to_yml(dir_name, package_path, yml_data, yml_text)
 
     with io.open(output_path, mode='w', encoding='utf-8') as f:
         f.write(yml_text)
-    return output_path, yml_path, script_path, image_path
+    return output_path, yml_path, script_path, image_path, desc_path
```
Comment (ronykoz) [inline package_creator.py]: Did you check you are not breaking anything this way?

File: package_creator.py [hunk 2]
```diff
@@ -82,6 +88,30 @@ def insert_image_to_yml(dir_name, package_path, yml_data, yml_text):
     return yml_text, found_img_path
 
 
+def insert_description_to_yml(dir_name, package_path, yml_data, yml_text):
```
Comment (ronykoz) [inline package_creator.py]: lets add:
if dir_name != "Integrations":
....

And remove it from line 107

File: package_creator.py [hunk 3]
```diff
@@ -40,7 +40,12 @@ def merge_script_package_to_yml(package_path, dir_name, dest_path=""):
     else:
         output_path = os.path.join(dir_name, output_filename)
 
-    yml_path = glob.glob(package_path + '*.yml')[0]
+    yml_paths = glob.glob(package_path + '*.yml')
+    yml_path = yml_paths[0]
+    for path in yml_paths:
```
[the remark this answers is not in the document]
Comment (liorblob) [reply package_creator.py]: According to the use case, the plugin saves a unified yaml file each time the integration/automation is exported/saved(to upload to Demisto for example). So I take the first yaml by default, then check if there's a file that's not unified.

File: package_creator.py (hunk 1 above)
Comment (liorblob) [reply package_creator.py]: What do you mean?

File: package_creator.py (hunk 2 above)
Comment (liorblob) [reply package_creator.py]: Ah, I forgot to dedup that

File: package_creator.py (hunk 1 above)
Comment (yaakovi) [reply package_creator.py]: check that no one is using this function outside the module (you changed the function interface)

**Comment by liorblob** [comment]
@ronykoz note the changes in `validate_files.py` and the new file `description.py` - take a look if you can

File: Tests/scripts/validate_files.py [hunk 4]
```diff
@@ -178,6 +183,10 @@ def validate_modified_files(self, modified_files):
                 if not image_validator.is_valid():
                     self._is_valid = False
 
+            description_validator = DescriptionValidator(file_path)
```
Comment (ronykoz) [inline Tests/scripts/validate_files.py]: why twice?

File: package_creator.py [hunk 5]
```diff
@@ -53,35 +63,58 @@ def merge_script_package_to_yml(package_path, dir_name, dest_path=""):
...
-        else:
-            yml_text = 'image: ' + image_data + '\n' + yml_text
-        # verify that our yml is good (loads and returns the image)
-        mod_yml_data = yaml.safe_load(yml_text)
-        yml_image = mod_yml_data.get('image')
-        assert yml_image.strip() == image_data.strip()
+    image_data, found_img_path = get_data(dir_name, package_path, "*png")
+    image_data = IMAGE_PREFIX + base64.b64encode(image_data)
+
+    if yml_data.get('image'):
```
Comment (ronykoz) [inline package_creator.py]: not needed if you made the hook fail in case the image is in the yml

File: package_creator.py [hunk 6]
```diff
@@ -112,24 +145,29 @@ def insert_script_to_yml(package_path, script_type, yml_text, dir_name, yml_data
 
     if dir_name == 'Scripts':
         if yml_data.get('script'):
-            yml_text = yml_text.replace(yml_data.get('script'), script_code)
-        else:
-            yml_text = yml_text.replace("script: ''", "script: " + script_code)
+            if yml_data['script'] != '-':
```
Comment (ronykoz) [inline package_creator.py]:
```suggestion
            if yml_data['script'] != '-' and yml_data['script'] != '':
```

File: package_creator.py [hunk 7]
```diff
@@ -112,24 +145,29 @@ def insert_script_to_yml(package_path, script_type, yml_text, dir_name, yml_data
...
+                print(yml_data['script'])
```
Comment (ronykoz) [inline package_creator.py]: remove

File: package_creator.py [hunk 8]
```diff
@@ -112,24 +145,29 @@ def insert_script_to_yml(package_path, script_type, yml_text, dir_name, yml_data
...
+                raise ValueError("Please change the script to be blank or a dash(-) for package {}"
+                                 .format(package_path))
 
     elif dir_name == 'Integrations':
         if yml_data.get('script', {}).get('script'):
-            yml_text = yml_text.replace(yml_data.get('script', {}).get('script'), script_code)
-        else:
-            yml_text = yml_text.replace("script: ''", "script: " + script_code)
-
+            if yml_data['script']['script'] != '-':
```
Comment (ronykoz) [inline package_creator.py]:
```suggestion
            if yml_data['script']['script'] != '-' and yml_data['script']['script'] != '-':
```

File: Tests/scripts/hook_validations/description.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import glob
+
+from Tests.test_utils import *
```
Comment (ronykoz) [inline Tests/scripts/hook_validations/description.py]: import only what you need

File: Tests/scripts/hook_validations/description.py [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+                                "please add a detailed description in order to proceed.")
+
+    def is_valid(self):
+        """Validate that the description exists."""
+        if self._is_valid is False:  # In case we encountered an IndexError in the init - we don't have a description
+            return self._is_valid
+
+        return self._is_valid
+
+    def is_existing_description(self):
```
Comment (ronykoz) [inline Tests/scripts/hook_validations/description.py]: you are not using it, should be called in the is_valid function

File: Tests/scripts/hook_validations/description.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+                print_error("You have added a detailed description in the yml "
+                            "file, please update the package {}".format(package_path))
+                return False
+
+            description_file = glob.glob(package_path + '/*.md')
+            if description_file:
+                is_description_in_package = True
+
+        if not (is_description_in_package or is_description_in_yml):
+            print_error("You have failed to add a detailed description in the package for {}".format(self.file_path))
```
Comment (ronykoz) [inline Tests/scripts/hook_validations/description.py]: `return False` here, and update the `self._is_valid = False`

File: Tests/scripts/hook_validations/description.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+
+        return is_description_in_package or is_description_in_yml
```
Comment (ronykoz) [inline Tests/scripts/hook_validations/description.py]: return True

File: Tests/scripts/hook_validations/description.py (hunk 9 above)
Comment (liorblob) [reply Tests/scripts/hook_validations/description.py]: That was copy paste :)

File: Tests/scripts/hook_validations/description.py [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        if '.md' not in self.file_path:
+            self.is_existing_description()
+
+        return self._is_valid
+
+    def is_existing_description(self):
+        """Check if the integration as a description."""
+        is_description_in_yml = False
+        is_description_in_package = False
+        if get_yaml(self.file_path).get('detaileddescription'):
```
Comment (ronykoz) [inline Tests/scripts/hook_validations/description.py]: might not be a yml

**Comment by liorblob** [comment]
@ronykoz can we merge?

File: Tests/scripts/hook_validations/description.py [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        self._is_valid = True
+
+        self.file_path = file_path
+
+    def is_valid(self):
+        self.is_existing_description()
+
+        return self._is_valid
+
+    def is_existing_description(self):
```
Comment (ronykoz) [inline Tests/scripts/hook_validations/description.py]: Now the name is no good, replace to `is_duplicate_description`

**Pull request merged**